### DISCUSSION TITLE
Is there an easier and cleaner way for Bit operations?

### DISCUSSION THREAD

**Discussion by kanimaru** [querydsl/querydsl#2804] [Q&A]
Hi Guys,

I try to implement tags system in native SQL with Querydsl sql and bit operators. Every bit is a tag.
But eighter JPA nor Hibernate supports them nativly so I try to go with Querydsl. I use postgres 11.x.
First I needed to override the generation of the domain object cause the column was bit and generator wanted to do this as boolean instead of a string. Postgres JDBC driver will return the values bit(32) column as string. That was not a big deal some custom type for hibernate and querydsl that converts string to bitset and vice versa. 

But now I struggle with the query api from querydsl.  
I mean this code works, but it feels strange and hope there is a cleaner way to do this:

```java
      var tagValue = Expressions.stringTemplate(String.format("'%s'::bit(32)", binaryTags));
      var tagContainsTemplate = Expressions.booleanTemplate("({0} & {1}) = {1}",  thing.tags, tagValue);

      var query = sqlQueryFactory.select(thing.all())
          .from(thing)
          .where(tagContainsTemplate);
```

First I tried something like this:

```java
      var tagValue = Expressions.stringTemplate("'{0}'::bit(32)", binaryTags);
      var tagContainsTemplate = Expressions.booleanTemplate("({0} & {1}) = {1}",  thing.tags, tagValue);

      var query = sqlQueryFactory.select(thing.all())
          .from(thing)
          .where(tagContainsTemplate);
```

but this results in a query that demands 2 parameters: 

```sql
select ... from things things where (things.tags & '?'::bit(32)) = '?'::bit(32)
```

with an exception: `The column index is out of range: 1, number of columns: 0.

Someone has a cleaner solution to this? Maybe custom operators in the configuration?

Best Regards 
Kani ~

**Comment by jwgmeligmeyling** [comment]
In templates you can use `{0s}` to indicate that the paramter should be constantified as literal. If you use `SQLQuery.setUseLiterals(true)` - then all constants are rendered as literal. In that case to create actual parameters, you need to instantiate a `new Param("name")` expression, and set it using `SQLQuery.set(param, value)`. By default, all constants are rendered as parameter instead of literal.
